Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
+    pb_istream_t* stream) {
+  pb_wire_type_t wire_type;
+  uint32_t tag;
+  bool eof;
+  bool status = pb_decode_tag(stream, &wire_type, &tag, &eof);
+  // TODO(rsgowman): figure out error handling: We can do better than a failed
+  // assertion.
+  FIREBASE_ASSERT(tag == google_firestore_v1beta1_MapValue_FieldsEntry_key_tag);
+  FIREBASE_ASSERT(wire_type == PB_WT_STRING);
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+  std::string key = DecodeString(stream);
+
+  status = pb_decode_tag(stream, &wire_type, &tag, &eof);
+  FIREBASE_ASSERT(tag ==
+                  google_firestore_v1beta1_MapValue_FieldsEntry_value_tag);
+  // NB: PB_WT_STRING is used for submessages too.
+  FIREBASE_ASSERT(wire_type == PB_WT_STRING);
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+
+  FieldValue value = DecodeSubFieldValue(stream);
+
+  return std::make_pair(key, value);
+}
+
+void EncodeObject(
+    pb_ostream_t* stream,
+    const std::map<const std::string, const FieldValue>& object_value) {
+  google_firestore_v1beta1_MapValue mapValue =
+      google_firestore_v1beta1_MapValue_init_zero;
+  // NB: c-style callbacks can't use *capturing* lambdas, so we'll pass in the
+  // object_value via the arg field (and therefore need to do a bunch of
+  // casting).
+  mapValue.fields.funcs.encode = [](pb_ostream_t* stream, const pb_field_t*,
+                                    void* const* arg) -> bool {
+    auto* object_value =
+        reinterpret_cast<const std::map<const std::string, const FieldValue>*>(
+            *arg);
+
+    // Encode each FieldEntry (i.e. key value pair.)
+    for (const auto& kv : *object_value) {
+      bool status =
+          pb_encode_tag(stream, PB_WT_STRING,
+                        google_firestore_v1beta1_MapValue_FieldsEntry_key_tag);
+      if (!status) {
+        // TODO(rsgowman): figure out error handling
+        abort();
+      }
+
+      EncodeFieldEntry(stream, kv);
+    }
+
+    return true;
+  };
+  mapValue.fields.arg =
+      const_cast<void*>(reinterpret_cast<const void*>(&object_value));
+
+  bool status = pb_encode_delimited(
+      stream, google_firestore_v1beta1_MapValue_fields, &mapValue);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+}
+
+std::map<const std::string, const FieldValue> DecodeObject(
+    pb_istream_t* stream) {
+  google_firestore_v1beta1_MapValue map_value =
+      google_firestore_v1beta1_MapValue_init_zero;
+  std::map<const std::string, const FieldValue> result;
+  // NB: c-style callbacks can't use *capturing* lambdas, so we'll pass in the
+  // object_value via the arg field (and therefore need to do a bunch of
+  // casting).
+  map_value.fields.funcs.decode = [](pb_istream_t* stream, const pb_field_t*,
+                                     void** arg) -> bool {
+    auto* result =
+        reinterpret_cast<std::map<const std::string, const FieldValue>*>(*arg);
+
+    std::pair<const std::string, const FieldValue> fv =
```

Comment:
Nit: I'd use `value_type` or `auto`.

Reply:
I don't like auto here (as it's unclear what type DecodeFieldEntry() returns without looking at it). I like FieldValue::ObjectT::value_type though (assuming we add that alias in.)

(deferred to chat).